modules/greenbidsBidAdapter.js

```diff
@@ -0,0 +1,240 @@
+import { getReferrerInfo, getPageTitle, getPageDescription, getConnectionDownLink } from '../libraries/pageInfosUtils/pageInfosUtils.js';
+/**
+ * @typedef {import('../src/adapters/bidderFactory.js').BidRequest} BidRequest
+ * @typedef {import('../src/adapters/bidderFactory.js').Bid} Bid
+ */
+
+const BIDDER_CODE = 'greenbids';
+const GVL_ID = 1232;
+const ENDPOINT_URL = 'https://hb.greenbids.ai';
+export const storage = getStorageManager({ bidderCode: BIDDER_CODE });
+
+export const spec = {
+  code: BIDDER_CODE,
+  gvlid: GVL_ID,
+  supportedMediaTypes: ['banner', 'video'],
+  /**
+   * Determines whether or not the given bid request is valid.
+   *
+   * @param {BidRequest} bid The bid params to validate.
+   * @return boolean True if this is a valid bid, and false otherwise.
+   */
+  isBidRequestValid: function (bid) {
+    if (typeof bid.params !== 'undefined' && parseInt(getValue(bid.params, 'placementId')) > 0) {
+      logInfo('Greenbids bidder adapter valid bid request');
+      return true;
+    } else {
+      logError('Greenbids bidder adapter requires placementId to be defined and a positive number');
+      return false;
+    }
+  },
+  /**
+   * Make a server request from the list of BidRequests.
+   *
+   * @param {validBidRequests[]} validBidRequests array of bids
+   * @param bidderRequest bidder request object
+   * @return ServerRequest Info describing the request to the server.
+   */
+  buildRequests: function (validBidRequests, bidderRequest) {
+    const bids = validBidRequests.map(bids => {
+      const reqObj = {};
+      let placementId = getValue(bids.params, 'placementId');
+      const gpid = deepAccess(bids, 'ortb2Imp.ext.gpid');
+      reqObj.sizes = getSizes(bids);
+      reqObj.bidId = getBidIdParameter('bidId', bids);
+      reqObj.bidderRequestId = getBidIdParameter('bidderRequestId', bids);
+      reqObj.placementId = parseInt(placementId, 10);
+      reqObj.adUnitCode = getBidIdParameter('adUnitCode', bids);
+      reqObj.transactionId = bids.ortb2Imp?.ext?.tid || '';
+      if (gpid) { reqObj.gpid = gpid; }
+    });
+    const topWindow = window.top;
+
+    const payload = {
+      referrer: getReferrerInfo(bidderRequest),
+      pageReferrer: document.referrer,
+      pageTitle: getPageTitle().slice(0, 300),
+      pageDescription: getPageDescription().slice(0, 300),
+      networkBandwidth: getConnectionDownLink(window.navigator),
+      timeToFirstByte: getTimeToFirstByte(window),
+      data: bids,
+      device: bidderRequest?.ortb2?.device || {},
+      deviceWidth: screen.width,
+      deviceHeight: screen.height,
+      devicePixelRatio: topWindow.devicePixelRatio,
+      screenOrientation: screen.orientation?.type,
+      historyLength: getHLen(),
+      viewportHeight: topWindow.visualViewport?.height,
+      viewportWidth: topWindow.visualViewport?.width,
+      hardwareConcurrency: getHC(),
+      deviceMemory: getDM(),
```

Comment:
we require that you add a disclosure to your documentation to use these api's

Reply:
see https://github.com/prebid/prebid.github.io/pull/5572

Reply:
and #12060

Reply:
I removed them, did not realised they would be for finger printing